subsys/net/lib/mqtt_sn/mqtt_sn.c

```diff
@@ -0,0 +1,1150 @@
+			return;
+		}
+
+		send_msg(client, msg);
+		mqtt_sn_msg_unref(msg);
+	}
+
+	if (client->evt_cb) {
+		client->evt_cb(client, &evt);
+	}
+}
+
+static void handle_puback(struct mqtt_sn_client *client, struct mqtt_sn_param_puback *p)
+{
+	struct mqtt_sn_publish *pub = mqtt_sn_publish_find_msg_id(client, p->msg_id);
+
+	if (!pub) {
+		LOG_ERR("No matching PUBLISH found for msg id %u", p->msg_id);
+		return;
+	}
+
+	mqtt_sn_publish_destroy(client, pub);
+}
+
+static void handle_pubrec(struct mqtt_sn_client *client, struct mqtt_sn_param_pubrec *p)
+{
+	struct mqtt_sn_msg *msg;
+	struct mqtt_sn_param response = { .type = MQTT_SN_MSG_TYPE_PUBREL };
+	struct mqtt_sn_publish *pub = mqtt_sn_publish_find_msg_id(client, p->msg_id);
+
+	if (!pub) {
+		LOG_ERR("No matching PUBLISH found for msg id %u", p->msg_id);
+		return;
+	}
+
+	pub->con.last_attempt = k_uptime_get();
+	pub->con.retries = N_RETRY;
+
+	response.params.pubrel.msg_id = p->msg_id;
+	msg = mqtt_sn_encode_msg(&response);
+	if (!msg) {
+		return;
+	}
+
+	send_msg(client, msg);
+	mqtt_sn_msg_unref(msg);
+}
+
+static void handle_pubrel(struct mqtt_sn_client *client, struct mqtt_sn_param_pubrel *p)
+{
+	struct mqtt_sn_msg *msg;
+	struct mqtt_sn_param response = { .type = MQTT_SN_MSG_TYPE_PUBCOMP };
+
+	response.params.pubcomp.msg_id = p->msg_id;
+	msg = mqtt_sn_encode_msg(&response);
+	if (!msg) {
+		return;
+	}
+
+	send_msg(client, msg);
+	mqtt_sn_msg_unref(msg);
+}
+
+static void handle_pubcomp(struct mqtt_sn_client *client, struct mqtt_sn_param_pubcomp *p)
+{
+	struct mqtt_sn_publish *pub = mqtt_sn_publish_find_msg_id(client, p->msg_id);
+
+	if (!pub) {
+		LOG_ERR("No matching PUBLISH found for msg id %u", p->msg_id);
+		return;
+	}
+
+	mqtt_sn_publish_destroy(client, pub);
+}
+
+static void handle_suback(struct mqtt_sn_client *client, struct mqtt_sn_param_suback *p)
+{
+	struct mqtt_sn_topic *topic = mqtt_sn_topic_find_msg_id(client, p->msg_id);
+
+	if (!topic) {
+		LOG_ERR("No matching SUBSCRIBE found for msg id %u", p->msg_id);
+		return;
+	}
+
+	if (p->ret_code == MQTT_SN_CODE_ACCEPTED) {
+		topic->state = MQTT_SN_TOPIC_STATE_SUBSCRIBED;
+		topic->topic_id = p->topic_id;
+		topic->qos = p->qos;
+	} else {
+		LOG_WRN("SUBACK with ret code %d", p->ret_code);
+	}
+}
+
+static void handle_unsuback(struct mqtt_sn_client *client, struct mqtt_sn_param_unsuback *p)
+{
+	struct mqtt_sn_topic *topic = mqtt_sn_topic_find_msg_id(client, p->msg_id);
+
+	if (!topic || topic->state != MQTT_SN_TOPIC_STATE_UNSUBSCRIBING) {
+		LOG_ERR("No matching UNSUBSCRIBE found for msg id %u", p->msg_id);
+		return;
+	}
+
+	mqtt_sn_topic_destroy(client, topic);
+}
+
+static void handle_pingreq(struct mqtt_sn_client *client)
+{
+	struct mqtt_sn_msg *msg;
+	struct mqtt_sn_param response = { .type = MQTT_SN_MSG_TYPE_PINGRESP };
+
+	msg = mqtt_sn_encode_msg(&response);
+	if (!msg) {
+		return;
+	}
+
+	send_msg(client, msg);
+	mqtt_sn_msg_unref(msg);
+}
+
+static void handle_pingresp(struct mqtt_sn_client *client)
```

Comment:
Could we drop the empty functions? I don't see a reason to introduce them, if particular replies don't require any processing.